### PR TITLE
s4513350 - Improved U-Net Model for ISIC Dataset

### PR DESCRIPTION
**Pull request by fzmi**
Hi Shakes and tutors,

This PR contains the implementation of the improved U-Net model for the ISIC skin defects segmentation. You can refer to README.md for more details and install instructions.

Main files added:
- `main.py` - The main driver script when running the project.
- `data.py` - Image data preprocessing.
- `metrics.py` - The loss functions for training the model.
- `model.py` - The implementation of the improved U-Net model.
- `plots.py` - Generate output images and metric plots.
- `demo.ipynb` - Additional Jupyter Notebook for displaying the training process step by step.

Thanks,
Yiyun Zhang
(45133507)

**19/11 Change request todos:**
- [x] Revert README.md changes
- [x] Report Dice Coefficient for the test set.

### CODE REVIEW

**Comment by fzmi** [comment]
Hi Shakes and tutors,

To add to this PR, you will notice a change to the `recognition/ISICs_Unet/README.md` file. This change is not part of my submission and has been existed as soon as you cloned the repository. The git seems bugged and cannot revert this file to its original state.

You can verify if it's the case by running:
- `git clone git@github.com:shakes76/PatternFlow.git`
- `cd ./PatternFlow`
- `git checkout topic-recognition`
- `git status`

Sample screenshot:
Screen Shot 2021-11-08 at 15 43 09

Thus, you can disregard any changes to the `recognition/ISICs_Unet/README.md` file. All my works are within the `recognition/Yiyun_ISIC` directory.

(This comment is added within the approved extension period.)

Thanks,
Yiyun Zhang
(45133507)

**Review by wooboyeong — CHANGES_REQUESTED** [review]
Please remove commit 3b8eb8b - not your README

**Comment by wooboyeong** [comment]
README
* Please report average DSC for test set in README

General Comments
* Code OK for EASY
* Commit messages OK

**Comment by fzmi** [comment]
Hi, thanks for reviewing my code. When I was investigating the causes that many students (including me) have unintentionally pushed an additional `README.md` file to their branches, I found out that this may be related to a Git case-sensitivity issue. On GitHub, there are apparently two directories with the same spellings but with different cases: `recognition/ISICs_UNet` and `recognition/ISICs_Unet`, which are poor naming practices made by other students. When I cloned the repository to my computer (mine is on macOS, not sure about Windows), Git recognised these two directories as the same one and merged all files into this folder, even with the settings `git config core.ignorecase` turned off. Since both original directories have the files with the same names `README.md`, this caused a file conflict and some common commands like `git restore` and `git clean` will simply not work.

**Relevant screenshots:**
On GitHub:
image

On local repository:
image

I have reverted the `README.md` by pushing a new commit since it is not a good practice to rewrite the commit history.

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present 
No Data leakage 
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No feedback required
Request Description good
